contracts/stake/src/state.rs

```diff
@@ -70,204 +70,219 @@ impl StakeState {
+                account: payload.account,
+                value: staked_value,
+                receiver: Some(transfer_payload.receiver),
             },
         );
 
-        let key = unstake.public_key.to_bytes();
+        let key = payload.account.to_bytes();
         self.previous_block_state
             .entry(key)
-            .or_insert((prev_value, unstake.public_key));
+            .or_insert((prev_stake, payload.account));
     }
 
     pub fn withdraw(&mut self, withdraw: Withdraw) {
-        // deplete the stake from a key and increment the signature counter
-        let loaded_stake = self
-            .get_stake_mut(&withdraw.public_key)
-            .expect("A stake should exist in the map to be withdrawn!");
+        self.check_new_block();
 
-        let counter = loaded_stake.counter();
-        let reward = loaded_stake.reward();
+        let payload = withdraw.payload();
+        let signature = withdraw.signature();
+        let transfer_payload = payload.transfer.payload();
 
-        if reward == 0 {
-            panic!("Nothing to withdraw!");
-        }
+        let loaded_stake = self
+            .get_stake_mut(&payload.account)
+            .expect("A stake should exist in the map to be unstaked!");
 
-        loaded_stake.deplete_reward();
-        loaded_stake.increment_counter();
+        let withdraw_value = transfer_payload.value();
 
-        // verify signature
-        let digest = Withdraw::signature_message(
-            counter,
-            withdraw.address,
-            withdraw.nonce,
-        )
-        .to_vec();
+        // ensure there is a reward, and that the withrawal not than the amount
+        // available
+        if withdraw_value > loaded_stake.reward {
+            panic!("Value withdrawn larger than available reward");
+        }
 
-        if !rusk_abi::verify_bls(
-            digest,
-            withdraw.public_key,
-            withdraw.signature,
-        ) {
+        // check signature is correct
+        let digest = payload.signature_message().to_vec();
+        if !rusk_abi::verify_bls(digest, payload.account, *signature) {
             panic!("Invalid signature!");
         }
 
-        // make call to transfer contract to mint the reward to the given
-        // address
-        let transfer_contract = TRANSFER_CONTRACT;
-        let _: bool = rusk_abi::call(
-            transfer_contract,
-            "mint",
-            &Mint {
-                address: withdraw.address,
-                value: reward,
-                sender: withdraw.public_key,
-            },
-        )
-        .expect("Minting a reward note should succeed");
+        // make call to transfer contract to withdraw a note from this contract
+        // containing
+        let _: () =
+            rusk_abi::call(TRANSFER_CONTRACT, "mint", &payload.transfer)
+                .expect("Withdrawing reward should succeed");
+
+        // update the state accordingly
+        loaded_stake.reward = 0;
```

Comment:
Well spotted! Nasty bug.

Reply:
Withdraw value cannot be less than reward, it will early panic at `value != loaded_stake.reward`

This is the main reason why the withdraw didn't have the amount specified, because it a provisioners is rewarded in the meanwhile it get stuck